doc/l2mc/IGMP_Snooping_HLD.md

```diff
@@ -0,0 +1,701 @@
+Refer to section 1.1
+
+# 3 Design
+## 3.1 Overview
+
+
+
+![IGMP Snooping](images/L2mc_Architecture_diagram.png "Figure 1: IGMP Snooping High Level Architecture")
+
+__Figure 1: IGMP Snooping High Level Architecture__
+
+L2MC docker will host L2mcastMgr and L2mcastd process.
+
+L2mcastMgr process will handle following interactions,
+
+-   L2mcastMgr subscribes to CONFIG_DB to get configuration changes, it gets VLAN info from STATE_DB, and STP state changes from APP_DB.
+-   L2mcastMgr notifies all these DB updates to L2mcastd for protocol operation via unix domain socket.
+
+L2mcastd process will handle following interactions,
+
+-   Linux kernel to processes dynamic IGMP and PIM hello packets and netlink events.
+-   L2mcastd uses libevent for processing below incoming events from l2mcastMgr,  
+    - CLI configurations  
+    - STP state change  
+    - VLAN configuration changes
+-   Timer events are generated every 100ms for handling IGMP protocol timers.
+-   L2mcastSync is part of L2mcastd which handles L2MC entries update to APP_DB to be consumed by L2mcOrch.
+
+
+## 3.2 DB Changes
+This section describes the changes made to different DBs for supporting IGMP Snooping.
+
+### 3.2.1 CONFIG DB
+#### CFG_L2MC_TABLE
+	
+    ;Store IGMP Snooping configuration per VLAN
+    ;Status: work in progress
+    key                         = CFG_L2MC_TABLE|"Vlan"vlanid
+    ;field                      = value
+    enabled                     = "true"/"false"                       ; Is the IGMP Snooping on VLAN enabled or disabled 
+    querier                     = "true"/"false"                       ; Is the IGMP Snooping querier enabled  or disabled 
+    fast-leave                  = "true"/"false"                       ; Is IGMP fast leave enabled or disabled 
+    version                     = 1*DIGIT                   ; IGMP Snooping Version ( Version 1 or 2 or 3, Default: 2)
+    query-interval              = 5*DIGIT                   ; Query Interval Value (1 to 18000sec, Default:125sec)
+    last-member-query-interval  = 5*DIGIT                   ; Last Member Query Interval Value (100 to 25500ms; Default: 1000ms)
+    query-max-response-time     = 2*DIGIT                   ; Query Response Time (1 to 25sec, Default: 10sec)
+     
+#### CFG_L2MC_MROUTER_TABLE
+    ;Store Mrouter configuration
+    ;Status: work in progress
+    key                         = CFG_L2MC_MROUTER_TABLE|"Vlan"vlanid|mrouter_port_name       
+    ;field                      = value               
+
+####  CFG_L2MC_STATIC_MEMBER_TABLE
+    ;Store Static L2MC entry configuration per VLAN. 
+    ;Status: work in progress 
+    key                 =  CFG_L2MC_STATIC_MEMBER_TABLE|"vlan"vlanid|group_address|static_port_name
+    ;field              = value
+
+### 3.2.2 APP DB
+
+#### APP_L2MC_VLAN_TABLE
+	;Store Snooping state on VLAN
+    ;Status: work in progress
+    key         = APP_L2MC_VLAN_TABLE:"Vlan"vlanid
+    ;field      = value
+
+#### APP_L2MC_MEMBER_TABLE
```

Comment:
Lets say for a given group 225.0.0.1, if there are two interested ports port 1(VLAN 10) and port 2(VLAN20), are we going to maintain two entries in the table? If this is the case, how are we going to check manage the SAI interactions where i guess we need to create mc_id for all the outgoing ports? does sync layer going maintain some data structure for entry and port list?

Reply:
As snooping works on Vlan, for the given scenario we will have db entry for both the Vlans like below,
APP_L2MC_MEMBER_TABLE:Vlan10:0.0.0.0:225.0.0.1:Port1
APP_L2MC_MEMBER_TABLE:Vlan20:0.0.0.0:225.0.0.1:Port2
Orchagent maintains data structure which holds (S, G, Vlan) entry and the interested ports for the multicast group on that Vlan. 
For each entry there will be mc_id associated, this mc_id will be updated with port when there is IGMP join/leave for multicast group.